app/src/main/java/com/capstone/event_finder/network/EventApi.java

```diff
@@ -0,0 +1,114 @@
+import com.parse.ParseUser;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
+
+public class EventApi {
+
+    private final EventRepository eventRepository;
+
+    public EventApi(EventRepository eventRepository) {
+        this.eventRepository = eventRepository;
+    }
+
+    public LiveData<List<Event>> getAPIEvents() {
+        String eventSearchRegion = "en_US";
+        Long eventSearchRadiusFromUserInMeters = 40000L;
+        String numberOfEventsToRetrieve = "10";
+        Long upcomingEventsOnly = (System.currentTimeMillis() / 1000L);
+        RetrofitClient.getInstance().getYelpAPI().getEvents(eventSearchRegion,
+                numberOfEventsToRetrieve,
+                upcomingEventsOnly,
+                eventSearchRadiusFromUserInMeters,
+                null,
+                ParseUser.getCurrentUser().getString("zip")).enqueue(new Callback<JsonObject>() {
+            @Override
+            public void onResponse(@NonNull Call<JsonObject> call, @NonNull Response<JsonObject> response) {
+                if (response.isSuccessful() && response.body() != null && !String.valueOf(convertToList(response.body())).equals("[]")) {
+                    addEventsToDatabase(response);
+                }
+            }
+
+            @Override
+            public void onFailure(@NonNull Call<JsonObject> call, @NonNull Throwable t) {
+                t.printStackTrace();
+            }
+        });
+        return null;
+    }
+
+    private Collection<? extends Event> convertToList(JsonObject result) {
+        List<Event> res = new ArrayList<>();
+        JsonArray events = result.getAsJsonArray("events");
+        for (int i = 0; i < events.size(); i++) {
+            JsonObject temp = (JsonObject) events.get(i);
+            Event event = new Event();
+            populateEventInfo(event, temp);
+            res.add(event);
+        }
+        return res;
+    }
+
+    public void populateEventInfo(Event event, JsonObject temp) {
+        event.setName(temp.get("name").getAsString());
+        event.setDescription(temp.get("description").getAsString());
+        event.setImageUrl(temp.get("image_url").getAsString());
+        event.setTimeStart(temp.get("time_start").getAsString());
+        event.setId(temp.get("id").getAsString());
+        event.setEventSiteUrl(temp.get("event_site_url").getAsString());
+        event.setCategory(temp.get("category").getAsString());
+        checkAndSetEventEndTime(event, temp);
+        checkAndSetEventCost(event, temp);
+        formatAndSetEventLocation(temp, event);
+    }
+
+    private void addEventsToDatabase(@NonNull Response<JsonObject> response) {
+        try {
+            JsonObject result = response.body();
+            assert result != null;
+            eventRepository.deleteAllEvents();
+            eventRepository.insert((List<Event>) convertToList(result));
```

Comment:
💡 We can introduce our _own_ callback pattern here, so we don't need to call `eventRepository` directly, much like Retrofit fits introduces a callback pattern for _us_.

Step 1:  Create an interface with the callback method.
```
public interface GetAPIEventsHandler {
   public void eventsRecieved(List<Event>);
}
```

Step 2:  Update our public method to take in this interface:
```
public LiveData<List<Event>> getAPIEvents(GetAPIEventsHandler eventHandler) {
```

Step 3:  Pass the handler down until we get to this method, and then replace these repository calls with:
```
eventHandler.eventsRecieved((List<Event>) convertToList(result));
```

Step 4:  Update any call to `getAPIEvents` to pass in our new handler:

```
eventApi.getAPIEvents(new GetAPIEventsHandler() {
   @Override 
   public void eventsRecieved(List<Event> events) {
        // do something with events, like save in the event repository
   }
});
```